### PR TITLE
Fix data race in fsm

### PR DESCRIPTION
**Pull request by 0xSasaPrsic**
# Description

Lock is added to prevent data race in hasQuorum method.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: consensus/polybft/consensus_runtime.go [hunk 1]
```diff
@@ -996,6 +998,9 @@ func (c *consensusRuntime) IsValidBlock(proposal []byte) bool {
 }
 
 func (c *consensusRuntime) IsValidSender(msg *proto.Message) bool {
+	c.lock.Lock()
```
Comment (JekaMas) [inline consensus/polybft/consensus_runtime.go]: is it Lock or Rlock?

Comment (0xSasaPrsic) [reply consensus/polybft/consensus_runtime.go]: good catch, I think it can be RL on all checks.

Comment (JekaMas) [reply consensus/polybft/consensus_runtime.go]: agree, only `c.fsm = fsm` at the very beginning of the PR should be a write lock.

**Review by JekaMas — APPROVED** [review]

**Review by JekaMas — APPROVED** [review]

File: consensus/polybft/consensus_runtime.go [hunk 2]
```diff
@@ -996,6 +998,9 @@ func (c *consensusRuntime) IsValidBlock(proposal []byte) bool {
 }
 
 func (c *consensusRuntime) IsValidSender(msg *proto.Message) bool {
+	c.lock.RLock()
```
Comment (igorcrevar) [inline consensus/polybft/consensus_runtime.go:1001]: I suggest doing on the top:
```
c.lock.RLock()
fsm := c.fsm
c.lock.RUnlock()
```
Or even better make `func getFsm` which will call every other function that needs fsm -> on the top
`fsm := c.getFsm()`

Comment (JekaMas) [reply consensus/polybft/consensus_runtime.go:1001]: getFSM() should do deep copying then. I believe it's a pointer type so just returning won't help much.

Comment (JekaMas) [reply consensus/polybft/consensus_runtime.go:1001]: @igorcrevar

**Review by b00ris — APPROVED** [review]

**Review by JekaMas — APPROVED** [review]

**Pull request merged**